### ISSUE TITLE
Implement Fast Fourier Transform ops

### ISSUE THREAD

**Issue by NickShahML** [tensorflow/tensorflow#386]
Hey TF,

Trying to integrate the new Unitary RNN's into tensorflow. Unfortunately, they require Fast Fourier Transform and Inverse Fast Fourier Transforms. 

For efficiency this needs to be done on the GPU. On the author's code (theano), they do this by making a Theano Op, and inserting scikit's Fast Fourier and Inverse Fast Fourier Here:

https://github.com/amarshah/complex_RNN/blob/master/fftconv.py

How can this be done in TF? Upon completing the Unitary RNN, I plan to share it to everyone! I asked on the google groups, but didn't get a reply.

**Comment by vrv** [comment]
We'll need to add a few FFT ops to get this done -- probably using cufft or fbfft for GPU.  Assigning to @zffchen78 since he has experience in ffts and adding ops :)

**Comment by LeavesBreathe** [comment]
Thank you @vrv and @zffchen78 . Basically, these new types of RNN's are very promising and I hope to integrate them into TensorFlow over the next week if possible. I'm starting to work on it here:
https://github.com/LeavesBreathe/Seq2Seq_Upgrade_TensorFlow

However, it will not be possible to do without FFT and IFFT support so I really appreciate your help!

**Comment by adder** [comment]
So would this imply that the FFT ops only work for GPU and not CPU? Or just very very slow on CPU?

**Comment by LeavesBreathe** [comment]
Well the problem is that the actual unitary RNN uses FFT and IFFT within its cell. Ideally you want the cell's computation done in parallel on the GPU. If its on the CPU, its going to take forever.

Here's the paper: Look at their main equation http://arxiv.org/pdf/1511.06464v1.pdf #6

**Comment by adder** [comment]
With 'taking forever', you mean compared to the uRNN GPU version or also compared to a 'classical' RNN or a LSTM with the same number of parameters? I would at least hope the speed to convergence demonstrated in the paper would outweigh any extra computation time per iteration. :)

Anyways, I would expect that all ops defined in Tensorflow would be available for both CPU and GPU (and any future device). Me and I guess others don't have day to day access to a suitable GPU. Or is that not one of the design goals?

**Comment by vrv** [comment]
We'll try to provide an implementation for both.

**Comment by LeavesBreathe** [comment]
Perfect! Thanks!

**Comment by zffchen78** [comment]
@adder , can you elaborate a bit more of what kind of FFT/IFFT you need? Specifically,
1) do you need r2c (real-to-complex), c2r(complex-to-real), or c2c(complex-complex) fft/ifft support?
2) do you need 1D, 2D, or 3D fft/ifft? or all of them?
3) do you need support for batching?
4) In your original email, you pointed to theano's code fftconv.py, which uses fft/ifft to implement conv2d operation. Do you need conv2d in the end? If that's all you want, tf has conv2d operation already and they run on gpu using cudnn (cudnn newer version uses fft under the hood, as I was told.)

**Comment by adder** [comment]
I guess you should ask this questions to @LeavesBreathe :)

**Comment by zffchen78** [comment]
@LeavesBreathe, is this the paper you are trying to reproduce? http://arxiv.org/pdf/1511.06464v1.pdf And formula (6) is the core computation you need to do. AFAIK, you need 2D, non-batched, complex-to-complex fft/ifft. Is that correct?

**Comment by LeavesBreathe** [comment]
@zffchen78, thanks for your help in this matter. I will do my best to answer your questions. The ultimate goal is to replicate this paper: http://arxiv.org/pdf/1511.06464v1.pdf 

To do, I was planning on making a new class in RNN_cell.py. Thereofre, I believe with need complex 2 complex along with 2d support. 

The reason why I wanted a pythonic tensorflow op is so that we can assign multiple gpu's to multiple unitary RNN's when the whole integration is complete. So one uRNN would be on gpu 1, and another uRNN would be on gpu 2. 

Whoops you beat me to it already!

**Comment by phsmit** [comment]
I would also very much like to see fft support, not for the training of dnn models, but for the feature extraction in a speech recognition pipeline. In this case only r2c, 1d would be required, preferably with batching both in cpu and gpu.

**Comment by LeavesBreathe** [comment]
Hey @zffchen78 , I just wanted to follow up and ask if this was implemented yet? I can't make any progress on the unitary RNN's without it. Don't mean to bring on any pressure! Just wanted to ask.

**Comment by zffchen78** [comment]
I implemented 2D fft in TF code base already. Hopefully, it'll be copied in OSS soon.

**Comment by LeavesBreathe** [comment]
Okay thanks for the headsup. Looking forward to it!

**Comment by phsmit** [comment]
@LeavesBreathe Why close this, it isn't there yet!

**Comment by zffchen78** [comment]
@LeavesBreathe, we pushed changes today which enables fft2d and ifft2d on gpu. You can take a look of python/kernel_tests/fft_ops_test.py to see if it works for you. We are still figuring out license issues w/ fftw where cpu supports of fft2d/ifft2d needs. Hopefully, gpu supports are sufficient for you now. Let me know if you see any problems.

**Comment by LeavesBreathe** [comment]
Awesome, gpu support is really all i needed so let me test it out and get back to you. I am going to be pretty busy over the holidays but come January 4 I should be back to testing this!

**Comment by zffchen78** [comment]
@LeavesBreathe, I took a brief of the paper you plan to replicate in TF. I suspect you'll encounter some nuances. E.g., some operations may not have support for complex64 as time being. They are naunces because they can be easily patched by updating a few .cc file to expand the types they support.

**Comment by LeavesBreathe** [comment]
OKay this is very good to know @zffchen78. I will be sure to look into these formats.

**Comment by raingo** [comment]
I think @LeavesBreathe answer to @zffchen78 question was wrong.

What uRNN needs is batched 1d fft. Both the paper and their theano implementation indicate this.

Paper wise, the Equation (6) is certainly multiplied with a complex vector for each data instance.

Implementation wise, check the following call stack,
1. https://github.com/amarshah/complex_RNN/blob/master/models.py#L14
2. https://github.com/amarshah/complex_RNN/blob/master/fftconv.py#L147

To TF core developer: why not simply implement an interface like scikits.cuda: https://github.com/lebedov/scikit-cuda/blob/master/skcuda/fft.py

**Comment by zffchen78** [comment]
We welcome the community to contribute the code.

If one wants to make the fft op supports more complete in tensorflow, feel
free to send us changes. It shouldn't be too hard to copy the fft2d impl,
and add fft1d, fft3d, and batched_fft1d, batched_fft2d, batched_3d, etc. as
needs arises, plus testing, etc. These ops' impl will pretty much look like
  stream->CreatePlan(...);
  stream->DoFft(...);

**Comment by rryan** [comment]
@raingo I added a batched 1D FFT (and 2D/3D) in https://github.com/tensorflow/tensorflow/commit/784778e76d3f5fd75b9ec99138770795907f1616.

**Comment by pputzky** [comment]
@rryan Thanks so much for your implementation. I tested your version of the batch_fft2, and I am afraid it does not return the expected results. There might be just a simple error in the transpose.

I have compared your batch version to the non batch version and numpy's fft2 in an ipython notebook. I have no idea how to upload the notebook here, so I have attached a PDF export of the notebook here
[test_batch_fft.pdf](https://github.com/tensorflow/tensorflow/files/185802/test_batch_fft.pdf).

What this shows is that your batch version of fft2 somehow seems to shuffle the rows and columns in the wrong way (that's my guess). If I compare TFs fft2 and numpy's fft2 they both give consistent results, however the batch_fft2 does not comply with the former two.

Good thing though: tf.batch_ifft2(tf.batch_fft2(x)) seems to return the identity.

**Comment by rryan** [comment]
Huh, interesting @pputzky -- the tests I added compare directly with numpy.
Could you please run tensorflow/python/kernel_tests/fft_ops_test.py and
share the result?

Also, what version of numpy and cuda (in particular cuFFT if you know) are
you using?

**Comment by rryan** [comment]
@pputzky -- ah, reading your notebook I think I see an issue:

```python
x_batch = tf.constant(mnist.test.images[1].reshape((1,28,28,1)), tf.float32)
x_fft_batch = batch_fourier(x_batch, (1,28,28,1))
```

```tf.fft2d``` transforms the inner-most 2 dimensions so that would be doing a 28x1 FFT2D. Can you try reshaping to (1, 28, 28)?

**Comment by pputzky** [comment]
@rryan Yes, that was precisely the problem. I only realised that after posting my comment, apologies for that. It works perfectly fine with  (1, 28, 28). Thanks again for the implementation!

**Comment by rryan** [comment]
@pputzky: :+1: glad to hear

**Comment by drufat** [comment]
Is there any update on FFT on the CPU? I need to be able to run it on a computer without a GPU, and currently it seems to be implemented on the GPU target.

**Comment by rryan** [comment]
@drufat unfortunately that's the case -- the current ops are GPU-only. 

If nobody from the TF team takes this on I would be up for doing it -- I think the most annoying part will be adding FFTW as a TF dependency.

**Comment by zffchen78** [comment]
We can't pull in FFTW because of its license implies complicated legal issues.

**Comment by bhack** [comment]
There is always [clFFT](https://github.com/clMathLibraries/clFFT/)

**Comment by zffchen78** [comment]
If your usage is not performance critical yet (if you do, probably you better get a gpu), you can always fall back to numpy for prototyping your model. E.g., here is how you can use numpy fft modules to in tf:

    # a bit excotic function (rfft)
    with self.test_session():
      x = tf.constant([1., 2., 3., 4.], tf.float32)
      def rfft(x):
        return np.fft.rfft(x).astype(np.complex64)
      y, = tf.py_func(rfft, [x], [tf.complex64])
      self.assertAllClose(y.eval(), np.fft.rfft([1., 2., 3., 4.]))

**Comment by alexhock** [comment]
I'm not really sure what's happening here but I'm not getting the performance from the gpu implementation of batch_fft2d that I was expecting. Somehow Numpy is much faster.

I'm using Tensorflow r0.8 and I've tried this with a gtx 980 on Fedora 23 and on a K40 on RedHat enterprise 7, CUDA 7.5 from NVIDIA. I'm using Anaconda 2.7.

Here is the code and results:

```python
import numpy as np
import tensorflow as tf
import time

x = np.ones((1, 4000, 4000), dtype=np.float32)
x = x * .1
e = tf.constant(x, tf.complex64)

with tf.Session() as sess:
    with sess.graph.device('/gpu:0'):
        print("starting fft")
        st = time.time()
        tf.batch_fft2d(e).eval()
        en = time.time()
        print ("gpu duration: {0}".format(en-st))


npst = time.time()
np.fft.fft2(x[0])
npen = time.time()

print ("npy duration: {0}".format(npen-npst))
```

The results are:
```
gpu duration: 14.0613360405 seconds
numpy duration: 1.05315804482 seconds
```
Am I missing something?

**Comment by rryan** [comment]
@alexhock -- I haven't tried your code myself but I believe you're hitting a common TensorFlow pitfall which is that you're also measuring the time TensorFlow spends constructing the graph and also the C++/Python conversion of the resulting tensor into a numpy array.

```python
g = tf.Graph()
x = np.ones((1, 4000, 4000), dtype=np.float32)
x = x * .1
with g.as_default():
  e = tf.constant(x, tf.complex64)
  fft = tf.batch_fft2d(e)

  # This barrier executes the FFT op as a dependency but does not pay the cost of converting the result of the FFT into Python objects.
  with tf.control_dependencies([fft]):
    barrier = tf.no_op(name='benchmark_barrier')

  with tf.Session() as sess:
    with tf.device('/gpu:0'):
      st = time.time()
      result = sess.run(barrier)
      en = time.time()
      print ("gpu duration: {0}".format(en-st))
```

Can you give that a try?

Also, note that to do a proper benchmark you're going to want to take multiple measurements in a loop and it's quite common to include a "warmup" phase of a dozen or so ```session.run``` calls that you ignore the results of.

**Comment by rryan** [comment]
Also note that Python garbage collection is going to be a source of noise in your benchmark (if you measure over thousands of runs) -- you can either use timeit [1] which turns off GC or you can use TensorFlow tracing [2] [3] to measure the op timing from within TensorFlow.

[1] https://docs.python.org/2/library/timeit.html
[2] https://www.tensorflow.org/versions/r0.8/api_docs/python/client.html#Session.run
[3] https://github.com/tensorflow/tensorflow/blob/fe454464681b036ff7fed3e42c6bb541fa52dd7c/tensorflow/core/protobuf/config.proto#L146

Pass a RunOptions proto with trace_level == FULL_TRACE and an empty RunMetadata proto to session.run and you'll get some tracing metrics for the run.

**Comment by zffchen78** [comment]
Ryan's right. Though a tiny correction (might not matter due to auto placement),  the code misplaced the with tf.device('/gpu:0') I think. It should be 

# Construct a computation graph and place a constant e and fft(e) on /gpu:0
with g.as_default(), tf.device('/gpu:0'):
  # Here we start to add nodes into 'g'
  e = tf.constant(x, tf.complex64)
  fft = tf.batch_fft2d(e)
  # Here we finish adding nodes into 'g'. 
  # Typically, a tf proram is easier to read and is more
  # robust to keep the graph g intact after this point.
 
  with tf.Session() as sess:
    for _ in range(100):
      sess.run(barrier)

tf.device('/gpu:0') has no bearing on how sess.run exectuting stuff. It affects how the computation graph is placed.

**Comment by rryan** [comment]
@zffchen78 --- oops, good catch. I was pre-☕. :)

**Comment by LeavesBreathe** [comment]
Thanks @rryan for implementing the batched form of the FFT and IFFT. Sorry I have been gone for a while, but I had other responsibilities I had to take care of.

I am working both on Unitary RNN's now along with Associated LSTM's -- both of which require complex number calculations. If I can get either of them working, I'll submit a push to Tensorflow's `rnn_cell.py`

Right now though, it doesn't seem like backprop is supported for complex numbers and I have inquired about it  #here:https://github.com/tensorflow/tensorflow/issues/2255

**Comment by el3ment** [comment]
Thanks @rryan! It would be great to also have real-valued FFT operations like batch_rfft2d() and batch_rifft2d() (which use half as many operations as the complex version). I think this is identical to the current implementation but using cufftExecR2C() and cufftExecC2R() instead of cufftExec()

**Comment by rryan** [comment]
Indeed! rfft is on my rainy-day TODO list. If somebody else would like to take that on before I get to it, please feel free.

**Comment by LeavesBreathe** [comment]
I'm working on an open source implementation of uRNN right now for everyone to use. I'm hoping with what is currently implemented, this is possible. Will post back in a few days with an update.

**Comment by jacksonloper** [comment]
Is there any update on FFT on the CPU?  Are fftw license issues still killing us?

**Comment by rryan** [comment]
I'm now working on a cuFFT-based rfft.

**Comment by rryan** [comment]
@jacksonloper CPU-based FFT is still in limbo unfortunately. I'll ping an internal thread about it.

**Comment by el3ment** [comment]
@rryan you are my hero.

**Comment by mborgerding** [comment]
If the FFTW licensing is an issue for the host FFT version, how about kissfft or the Eigen version thereof?

**Comment by mwoodson1** [comment]
Any update on the rfft? If not I can possibly pick up some work on that.

**Comment by rryan** [comment]
Sorry for the delay -- I have a working 1D and 2D (forward and backward) implementation. I'm working on 3D but haven't gotten the gradients correct yet. I hope to get this done soon.

**Comment by beniroquai** [comment]
@rryan Are you talking about the general fft2d or the rfft in 1D and 2D?  It would be really nice to have all of them in the CPU version as well. Thank you for your help :)

**Comment by rryan** [comment]
@mwoodson1 RFFT/IRFFT (1D/2D/3D with gradients for 1D/2D) is now committed internally, once the GitHub sync occurs it'll be available.

Everything is still implemented with cuFFT, so it's GPU-only unfortunately. We're still working on a solution for CPU FFTs.

**Comment by Timeroot** [comment]
Is there a workaround for CPU-based ffts? Any kind, really (real2complex, complex2complex, whatever), I'd just like to be able to do prototyping when I'm without my GPU access. :)

**Comment by rryan** [comment]
@Timeroot as a workaround until we get CPU FFT kernels added you can compute a 1D Fourier transform on CPUs with a matrix multiply:
```python
complex_signal = ...
ft_length = 1024
# Compute the DFT matrix for a given length. scipy has a built-in for this.
dft_matrix = tf.constant(scipy.linalg.dft(ft_length).astype(np.complex64))
c2c_result = tf.matmul(complex_signal, dft_matrix)
```

If you want the r2c, just slice the `fft_length / 2 + 1` unique components.

```python
complex_signal = tf.complex(real_signal, tf.zeros_like(real_signal))
... example as above ...
unique = ft_length / 2 + 1
r2c_result = c2c_result[..., :unique]
```

**Comment by tillahoffmann** [comment]
@rryan, I have working CPU/GPU kernels for complex FFTs based on Eigen but have some issues integrating your recent RFFT changes. Would you be up for a chat to iron out those bumps so we can add the CPU kernels? Ideally, I would like to move to an interface similar to `conv2d` which takes `use_cudnn_on_gpu` as an optional argument.

**Comment by rryan** [comment]
@tillahoffmann happy to chat! though I'm not on the TF team per se (just a contributor). I'll loop in some TF folks who could weigh in. Want to email me and we'll go from there? My google.com email is rjryan. Also, are the changes pushed publicly anywhere?

**Comment by tillahoffmann** [comment]
@rryan, no not pushed publicly at the moment because I don't have the tests passing yet. Will drop you an e-mail.

**Comment by rryan** [comment]
The current plan of record is that @tillahoffmann is working on a CPU-based FFT/RFFT using Eigen's TensorFFT module. No firm ETA but hopefully soon -- thanks @tillahoffmann!

**Comment by pawarrick** [comment]
@tillahoffmann, do you have any updates on progress to the CPU implementation of FFT?  Many thanks for this.

**Comment by tillahoffmann** [comment]
@pawarrick, complex FFT and IFFT as well as the RFFT are now in master. The IRFFT still needs a bit of work.

**Comment by pawarrick** [comment]
yes thanks very much for your contribution. I saw the discussion about it in this thread #9029

**Comment by Czxck001** [comment]
@tillahoffmann Very much thanks for your contribution! I've made up the IRFFT part, see #10127

**Comment by rryan** [comment]
Thanks to @tillahoffmann (#9029) and @Czxck001 (#10127), I think we can call this done!

**Comment by beniroquai** [comment]
Great work! Now can I finally hack any fft operation on my laptop. 
I'm a bit curious. Do you see any chances, that these OPs could eventually be ported to mobile platforms such as Android?

**Comment by Androbin** [comment]
@beniroquai 
As long as they are available for CPU, they can be compiled for Android.
I don't know, if they are in the default op list.
Otherwise you can use selective registration.

**Comment by uschmidt83** [comment]
Hi, I just tried running FFT ops on the CPU with the 1.2.1 release, but still get errors like this:
`No OpKernel was registered to support Op 'FFT2D' with these attrs.  Registered devices: [CPU], Registered kernels:
  device='GPU'`

Am I missing something?

**Comment by tillahoffmann** [comment]
@uschmidt83, have you tried the [nightly build](https://github.com/tensorflow/tensorflow#installation)?

**Comment by beniroquai** [comment]
Is it scheduled to put it also in the release version at one point?

**Comment by rryan** [comment]
It should be in TF 1.3 which is in the process of being released! Also note
tf.contrib.signal in this release which has some signal processing basics
that are useful for e.g. computing spectrograms.

**Comment by uschmidt83** [comment]
Thanks, everyone! It does work for me with the 1.3.0-rc0 release.

**Comment by FedericoMuciaccia** [comment]
there isn't an alias for real ffts in TF 1.3.0

present functions:
`tf.spectral.fft`
`tf.spectral.fft2d`
`tf.spectral.fft3d`
`tf.spectral.ifft`
`tf.spectral.ifft2d`
`tf.spectral.ifft3d`
`tf.spectral.irfft`
`tf.spectral.irfft2d`
`tf.spectral.irfft3d`
`tf.spectral.rfft`
`tf.spectral.rfft2d`
`tf.spectral.rfft3d`

present aliases:
`tf.fft`
`tf.fft2d`
`tf.fft3d`
`tf.ifft`
`tf.ifft2d`
`tf.ifft3d`

can someone please create those missing aliases for the next TF release?

**Comment by rryan** [comment]
Hi Federico,

This is intentional as we are trying to avoid polluting the top-level
namespace in favor of more specific submodules. tf.fft and other fft
symbols in the tf module are deprecated now :).

**Comment by WangNuoWa** [comment]
Hi  everyone,

I would like to do some work on the basis of this article, but the original author uses theano.tensor.fft module to implement batch 1-D FFT, batch 2-D FFT, batch 1-D IFFT, batch 2-D IFFT. Now i want to replace the theano.tensor.fft module directly into tensorflow in tf.fft and tf.fft2d, ok?

<https://github.com/ChihebTrabelsi/deep_complex_networks/blob/master/complexnn/fft.py>

We look forward to your help, thank you very much!

**Comment by rryan** [comment]
@WangNuoWa -- yes, tf.fft/tf.ifft and tf.fft2d/tf.ifft2d should be drop-in replacements for those (though I have never used theano.tensor.fft), perhaps with slightly different normalization strategies though (I see an "ortho" norm in there).

**Comment by sirgogo** [comment]
Is tf.fft differentiable?

**Comment by rryan** [comment]
Yes!

**Comment by sirgogo** [comment]
@rryan, is there anything special we would have to do? I get new warnings now, like these: (you can copy paste the whole thing)

```
import numpy as np
import tensorflow as tf

x = np.linspace(0,10,num=10,dtype=np.complex64)
w = np.linspace(0,1,num=10,dtype=np.complex64)
y = np.multiply(x,w)

x_tf = tf.Variable(x,dtype=tf.complex64)
w_tf = tf.Variable(w,dtype=tf.complex64)
y_tf = tf.multiply(x_tf,w_tf)

sess = tf.Session()
sess.run(tf.global_variables_initializer())
sess.run(y_tf)
g1 = tf.gradients(y_tf,x_tf)
'''
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/intel/intelpython3/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py", line 472, in gradients
    grad_ys = _DefaultGradYs(grad_ys, ys, colocate_gradients_with_ops)
  File "/opt/intel/intelpython3/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py", line 233, in _DefaultGradYs
    y.dtype)
TypeError: Gradients of complex tensors must set grad_ys (y.dtype = tf.complex64)
'''

a_tf = tf.fft(x_tf)
sess.run(a_tf)
g2 = tf.gradients(a_tf,x_tf)
'''
g2 = tf.gradients(a_tf,x_tf)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/intel/intelpython3/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py", line 472, in gradients
    grad_ys = _DefaultGradYs(grad_ys, ys, colocate_gradients_with_ops)
  File "/opt/intel/intelpython3/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py", line 233, in _DefaultGradYs
    y.dtype)
TypeError: Gradients of complex tensors must set grad_ys (y.dtype = tf.complex64)
'''

print(x_tf)
print(w_tf)
print(y_tf)
'''
>>> print(x_tf)
<tf.Variable 'Variable:0' shape=(10,) dtype=complex64_ref>
>>> print(w_tf)
<tf.Variable 'Variable_1:0' shape=(10,) dtype=complex64_ref>
>>> print(y_tf)
Tensor("Mul:0", shape=(10,), dtype=complex64)
'''
```

**Comment by rryan** [comment]
> @rryan, is there anything special we would have to do? I get new warnings now, like these: (you can copy paste the whole thing)

I think what you're seeing is independent of whether FFT is differentiable. [Here's a colab](
https://colab.research.google.com/notebook#fileId=1fh1D_okZ8fRLOyjHRiaY7hOS5bBHfOIy) where I copied your example. You need to do as the error message says and provide a `grad_ys` argument to `tf.gradients` if `ys` is complex, because the TF gradient infrastructure currently won't assume that `grad_ys` is all ones if the type is complex, as it does when the type is real (I'm not totally sure why this is... maybe the original author of that code didn't want to assume that `1+1j` was the proper `grad_ys` for a complex number because its magnitude is greater than 1?). 

You can see this logic here:
https://github.com/tensorflow/tensorflow/blob/70e33e0f77cb4f92ddd70a8dd601873b178124fc/tensorflow/python/ops/gradients_impl.py#L232-L239

**Comment by zaccharieramzi** [comment]
Hi,

I was wondering whether you guys at TF were planning to implement a non-uniform FFT layer.
I don't know if this question belongs here, or if I should create a new issue by itself.

Cheers

**Comment by rryan** [comment]
@zaccharieramzi nobody on the TF team is currently considering that, as far as I know. 

AIUI, the most efficient algorithms for it are defined in terms of the FFT -- is it an easy extension of the FFT ops that already exist? If so, I would be happy to code review an addition to tf.signal. to add it if someone submits a pull request. 

If it is more involved to implement, for example requiring custom kernels, then I think it might be a harder sell to include since there has not been high demand for it, and adding custom kernels increasing the binary size and compile time of TensorFlow.  

Could you please file a GitHub issue requesting the feature? Anyone else who is interested would then have a nice place to add a +1 so we can better understand how many people would benefit from it being in TensorFlow itself (as opposed to being implemented as a standalone library).

**Comment by zaccharieramzi** [comment]
Hi @rryan ,

Unfortunately it's not a direct extension of the FFT afaik (I am still new to the topic). So I think it does require a custom kernel. I totally understand that it could be a hard sell and I will file a github issue requesting the feature (I am totally not familiar with C/C++ so I don't think I could implement it myself).

In the meantime I will use [ODL](https://github.com/odlgroup/odl) to achieve my goals, although slower because the implementation is in python.

**Comment by itskobold** [comment]
Hello,

I'm getting this not implemented error...

`Gradient for IRFFT3D is not implemented.`  (I'm assuming I'll get a similar error for RFFT3D too).

Is there any way I could get this working myself?  I've taken a look at where the ops for 1D/2D (I)RFFT are defined in fft_ops.py but I'm a bit lost as to what I'd need to do to implement it in 3D.

Thanks in advance for any advice!

**Comment by tillahoffmann** [comment]
I don't have tensorflow installed at the moment to verify, but the three-dimensional RFFT is just a two-dimensional RFFT plus a FFT along the leading dimension. Here's a numpy example.

```python
import numpy as np

x = np.random.normal(0, 1, (15, 16, 17))
# Compute three-dimensional FFT.
fftn = np.fft.rfftn(x, x.shape)
# Do it in two steps.
fftn_two_step = np.fft.fft(np.fft.rfft2(x), axis=0)
np.testing.assert_allclose(fftn, fftn_two_step)
```

This should allow you to get the gradients by decomposing the three-dimensional transform into a one-dimensional and a two-dimensional transform (assuming those two have gradients implemented).
